src/Common/src/Interop/Windows/NCrypt/Interop.NCryptDeriveKeyMaterial.cs

```diff
@@ -0,0 +1,258 @@
+
+                // Figure out how big the key material is
+                ErrorCode error = NCryptDeriveKey(
+                    secretAgreement,
+                    kdf,
+                    ref parameterDesc,
+                    null,
+                    0,
+                    out int keySize,
+                    flags);
+
+                if (error != ErrorCode.ERROR_SUCCESS && error != ErrorCode.NTE_BUFFER_TOO_SMALL)
+                {
+                    throw new CryptographicException((int)error);
+                }
+
+                // Allocate memory for the key material and generate it
+                byte[] keyMaterial = new byte[keySize];
+
+                error = NCryptDeriveKey(
+                    secretAgreement,
+                    kdf,
+                    ref parameterDesc,
+                    keyMaterial,
+                    keyMaterial.Length,
+                    out keySize,
+                    flags);
+
+                if (error != ErrorCode.ERROR_SUCCESS)
+                {
+                    throw new CryptographicException((int)error);
+                }
+
+                return keyMaterial;
+            }
+        }
+
+        /// <summary>
+        ///     Derive key material from a secret agreement using a hash KDF
+        /// </summary>
+        internal static byte[] DeriveKeyMaterialHash(
+            SafeNCryptSecretHandle secretAgreement,
+            string hashAlgorithm,
+            byte[] secretPrepend,
+            byte[] secretAppend,
+            SecretAgreementFlags flags)
+        {
+            return DeriveKeyMaterial(
+                secretAgreement,
+                BCryptNative.KeyDerivationFunction.Hash,
+                hashAlgorithm,
+                null,
+                secretPrepend,
+                secretAppend,
+                flags);
+        }
+
+        /// <summary>
+        ///     Derive key material from a secret agreement using a HMAC KDF
+        /// </summary>
+        internal static byte[] DeriveKeyMaterialHmac(
+            SafeNCryptSecretHandle secretAgreement,
+            string hashAlgorithm,
+            byte[] hmacKey,
+            byte[] secretPrepend,
+            byte[] secretAppend,
+            SecretAgreementFlags flags)
+        {
+            return DeriveKeyMaterial(
+                secretAgreement,
+                BCryptNative.KeyDerivationFunction.Hmac,
+                hashAlgorithm,
+                hmacKey,
+                secretPrepend,
+                secretAppend,
+                flags);
+        }
+
+        /// <summary>
+        ///     Derive key material from a secret agreeement using the TLS KDF
+        /// </summary>
+        internal static byte[] DeriveKeyMaterialTls(
+            SafeNCryptSecretHandle secretAgreement,
+            byte[] label,
+            byte[] seed,
+            SecretAgreementFlags flags)
+        {
+            NCryptBuffer[] buffers = new NCryptBuffer[2];
```

Comment:
In a previous method these were stackallocated... can that be done here, too?

Reply:
Good catch, done.  (Good thing I fixed the span even though I "knew" the only caller (singular) was stackalloced :))